[CIR] Add support for thread-local storage (TLS)

adams381 opened the pull request:
This commit adds full support for thread-local storage variables in ClangIR, including code generation, lowering to LLVM IR, and comprehensive testing.

Changes include:
- Added CIR_TLSModel enum with 4 TLS models (GeneralDynamic, LocalDynamic, InitialExec, LocalExec) to CIROps.td
- Extended GlobalOp with optional tls_model attribute
- Extended GetGlobalOp with thread_local unit attribute
- Added verification to ensure thread_local GetGlobalOp references globals with tls_model set
- Implemented GetDefaultCIRTLSModel() and setTLSMode() in CIRGenModule
- Updated getAddrOfGlobalVar() to handle TLS access
- Removed MissingFeatures assertions for TLS operations
- Added lowering of GetGlobalOp with TLS to llvm.threadlocal.address intrinsic
- Added lowering of GlobalOp with tls_model to LLVM thread_local globals
- Added comprehensive test with CIR, LLVM, and OGCG checks

Known limitations (matching incubator):
- Static local TLS variables not yet implemented
- TLS_Dynamic with wrapper functions not yet implemented

Fixes #153270

adams381 committed:
[CIR] Add support for thread-local storage (TLS)

This commit adds full support for thread-local storage variables in ClangIR,
including code generation, lowering to LLVM IR, and comprehensive testing.

Changes include:
- Added CIR_TLSModel enum with 4 TLS models (GeneralDynamic, LocalDynamic,
  InitialExec, LocalExec) to CIROps.td
- Extended GlobalOp with optional tls_model attribute
- Extended GetGlobalOp with thread_local unit attribute
- Added verification to ensure thread_local GetGlobalOp references globals
  with tls_model set
- Implemented GetDefaultCIRTLSModel() and setTLSMode() in CIRGenModule
- Updated getAddrOfGlobalVar() to handle TLS access
- Removed MissingFeatures assertions for TLS operations
- Added lowering of GetGlobalOp with TLS to llvm.threadlocal.address intrinsic
- Added lowering of GlobalOp with tls_model to LLVM thread_local globals
- Added comprehensive test with CIR, LLVM, and OGCG checks

Known limitations (matching incubator):
- Static local TLS variables not yet implemented
- TLS_Dynamic with wrapper functions not yet implemented

This fixes issue #153270

xlauko commented on clang/include/clang/CIR/Dialect/IR/CIROps.td:
```diff
@@ -1988,16 +2073,22 @@ def CIR_GetGlobalOp : CIR_Op<"get_global", [
   let results = (outs Res<CIR_PointerType, "", []>:$addr);
 
   let assemblyFormat = [{
+    (`thread_local` $tls^)?
```
```suggestion
    (`thread_local` $thread_local^)?
```

bcardosolopes commented:
Thanks for working on this. This PR handles both TLS and `BinOpOverflowOp`, they seem orthogonal and can be done incrementally - should be split in multiple PRs.

andykaylor commented on clang/lib/CIR/CodeGen/CIRGenModule.cpp:
```diff
@@ -1933,6 +1902,35 @@ void CIRGenModule::setGVPropertiesAux(mlir::Operation *op,
+
+  // Override the TLS model if it is explicitly specified.
+  if (d.getAttr<TLSModelAttr>()) {
+    llvm_unreachable("NYI");
```
```suggestion
    errorNYI(d.getSourceRange(), "TLS model attribute");
```

andykaylor commented on clang/lib/CIR/CodeGen/CIRGenModule.cpp:
```diff
@@ -1933,6 +1902,35 @@ void CIRGenModule::setGVPropertiesAux(mlir::Operation *op,
+void CIRGenModule::setTLSMode(mlir::Operation *op, const VarDecl &d) const {
+  assert(d.getTLSKind() && "setting TLS mode on non-TLS var!");
+
+  auto tlm = GetDefaultCIRTLSModel();
```
```suggestion
  cir::TLS_Model tlm = getDefaultCIRTLSModel();
```
Even though the function definition is right above here currently, I don't think we want to use `auto` here.

andykaylor reviewed (commented):
This looks great. I have a bunch of small requests, but overall this is good.

adams381 committed:
Remove unnecessary intermediate variables in GetGlobalOp usage

Simplified code by directly returning the result of GetGlobalOp::create()
instead of storing it in an intermediate variable and calling getAddr().
The create method returns a value that implicitly converts to mlir::Value,
so the intermediate step is unnecessary.

This addresses reviewer feedback.

adams381 committed:
Address reviewer feedback for TLS implementation

Changes based on code review feedback:
- Use cast<> instead of dyn_cast<> + assert for type safety
- Replace llvm_unreachable with errorNYI for better diagnostics in static var decl
- Fix function name capitalization: GetDefaultCIRTLSModel -> getDefaultCIRTLSModel
- Remove auto where type is not obvious (tlm variable)
- Remove const from setTLSMode to allow errorNYI call
- Add invalid CIR test to verify TLS verification logic

These changes improve code quality and follow LLVM coding standards.

andykaylor reviewed (approved):
lgtm

andykaylor commented on clang/lib/CIR/CodeGen/CIRGenModule.cpp:
```diff
@@ -1933,6 +1900,34 @@ void CIRGenModule::setGVPropertiesAux(mlir::Operation *op,
+  cir::TLS_Model tlm = getDefaultCIRTLSModel();
+
+  // Override the TLS model if it is explicitly specified.
+  if (d.getAttr<TLSModelAttr>()) {
```
I was going to just let this go, but since this needs to be rebased anyway I'll mention it. This shouldn't have braces since the body is a single line.

https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements

adams381 committed:
Remove braces from single-line if statement

Address reviewer feedback from PR #168662. Remove unnecessary
braces from single-statement if body to follow LLVM coding
standards.

adams381 commented:
Addressed reviewer feedback: Removed unnecessary braces from single-line if statement to follow LLVM coding standards per https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements

Ready for another review!

andykaylor reviewed (approved):
This looks good, but it needs to be rebased.

adams381 committed:
Update TLS test to match upstream function attribute changes

The function now includes no_inline attribute due to upstream
main branch changes. Update test expectation accordingly.

adams381 committed:
Fix formatting in CIRBaseBuilder.h

Fix clang-format violations in createGetGlobal method.
